test/src/input_validation/function_test.dart

```diff
@@ -37,128 +37,141 @@ void main(){
         expect(tester.testFunction(string), equals(false));
       });
 
       test('invalid comma position 4', () {
-        var string = 'max ( , 4 )';
+        var string = 'max(,4)';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('invalid comma function', () {
-        var string = 'sin ( 4 , 2 )';
+        var string = 'sin(4,2 )';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('invalid complex comma function', () {
-        var string = 'tan ( 2 + sin ( 2 ) , ( 3 * 4 ) * cos ( 2 ) )';
+        var string = 'tan(2+sin(2),(3*4)*cos(2))';
         expect(tester.testFunction(string), equals(false));
       });
     });
     
     group('no operands next to one another', () {
       test('no plus adjacent', () {
-        var string = '2 + + 2';
+        var string = '2++2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no multiplication adjacent', () {
-        var string = '2 * * 2';
+        var string = '2**2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no division adjacent', () {
-        var string = '2 / / 2';
+        var string = '2//2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no subtraction adjacent', () {
-        var string = '2 - - 2';
+        var string = '2--2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no double exponentiation', () {
-        var string = '2 ^ ^ 2';
+        var string = '2^^2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no plus adj div', () {
-        var string = '2 + / 2';
+        var string = '2+/2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no multiply adj plus', () {
-        var string = '2 * + 2';
+        var string = '2*+2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no div adj multiply', () {
-        var string = '2 / * 2';
+        var string = '2/*2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no div adj plus', () {
-        var string = '2 / + 2';
+        var string = '2/+2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no sub adj plus', () {
-        var string = '2 - + 2';
+        var string = '2-+2';
         expect(tester.testFunction(string), equals(false));
       });
 
       test('no exp adj div', () {
-        var string = '2 ^ / 2';
+        var string = '2^/2';
         expect(tester.testFunction(string), equals(false));
       });
     });
 
     group('invalid pairs of parentheses', () {
       test('no unmatched open', () {
-        var string = '2 * ( ( 2 )';
-        expect(tester.testFunction(string), equals(false));
+        var string = '2*((2)';
+        expect(tester.testFunction(string), isTrue);
```

Comment:
should be 'equals(false)' instead of isTrue

Reply:
this works on a calculator and also works on ours

Reply:
I see, now, we're taking the input directly from the user and the translator would add the last ) automatically making the expression valid. That is why we can make this case valid.